File: rec-core/src/main/java/net/kimleo/rec/rule/impl/Exist.java
package net.kimleo.rec.rule.impl;

import net.kimleo.rec.Pair;
import net.kimleo.rec.record.Record;
import net.kimleo.rec.repository.RecordSet;
import net.kimleo.rec.rule.RecRule;
import net.kimleo.rec.rule.Result;

import java.util.ArrayList;
import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;

public class Exist implements RecRule {

    @Override
    public Pair<Boolean, List<Result>> verify(List<RecordSet> recs) {
        assert recs.size() == 2;

        boolean successful = true;
        ArrayList<Result> result = new ArrayList<>();

        RecordSet image = recs.get(0);
        RecordSet complete = recs.get(1);

        Set<Record> set = complete.getRecords().collect(Collectors.toSet());

        for (Record rec: image) {
            if (!set.contains(rec)) {
                result.add(() -> String.format("%s of %s cannot be found in %s",
                        rec.getText(),
                        image.getConfig().name(),
                        complete.getConfig().name()));
                successful = false;
            }
        }

        return new Pair<>(successful, result);
    }
}

File: rec-core/src/main/kotlin/net/kimleo/rec/rule/impl/Exist.kt
package net.kimleo.rec.rule.impl

import net.kimleo.rec.repository.RecordSet
import net.kimleo.rec.rule.RecRule
import net.kimleo.rec.rule.Result

class Exist: RecRule {
    override fun verify(recs: List<RecordSet>): Pair<Boolean, List<Result>> {
        assert(recs.size == 2)

        var successful = true
        val result = arrayListOf<Result>()

        val image = recs[0]
        val complete = recs[1]

        val complete_set = complete.toSet()


        for (el in image) {
            if (!complete_set.contains(el)) {
                successful = false
                result.add(object: Result {
                    override val details = "${el.text} of ${image.config.name} cannot be found in ${complete.config.name}"
                })
            }
        }

        return Pair(successful, result)
    }
}